File: src-javascript/engine/utils/math2d.js
"use strict";

const MATH2D_DEG_TO_RAD = Math.PI / 180.0;
const MATH2D_MAX_INT32 = 0x7FFFFFFF;
const MATH2D_LOG100 = Math.log(100);
const MATH2D_HALF_PI = Math.PI / 2.0;
const MATH2D_PRNG = new Uint32Array(1);

function math2d_random(min, max) {
    //return Math.random() * (max - min + 1) + min;
    window.crypto.getRandomValues(MATH2D_PRNG);
    let pecent = MATH2D_PRNG[0] / 0xffffffff;
    return pecent * (max - min + 1) + min;
}

function math2d_random_int(min, max) {
    let value = math2d_random(min, max);
    value = Math.trunc(value);

    if (value < min) return min;
    if (value > max) return max;
    return value;
}


function math2d_lerp(start, end, step) {
    return start + (end - start) * step;
}

function math2d_inverselerp(start, end, value) {
    return (value - start) / (end - start);
}

function math2d_nearestdown(value, step) {
    return value - (value % step);
}

function math2d_cubicbezier(offset, point0, point1, point2, point3) {
    let neg = 1 - offset;
    return (neg * neg * neg) * point0 +
        3 * offset * (neg * neg) * point1 +
        3 * (offset * offset) * neg * point2 +
        (offset * offset * offset) * point3;
}

function math2d_clamp(value, min, max) {
    if (value < min) return min;
    if (value > max) return max;
    return value;
}

function math2d_rotate_point(point, radians) {
    let s = Math.sin(radians);
    let c = Math.cos(radians);
    let x = point[0] * c - point[1] * s;
    let y = point[0] * s + point[1] * c;

    point[0] = x;
    point[1] = y;

    return point;
}

function math2d_color_blend_normal(back_color, front_color, output_color) {
    //
    // Avoid useless calculations if both colors are equal, also both alphas must be 1.0
    //
    L_check_equal: {
        for (let i = 0; i < 4; i++) {
            if (back_color[i] != front_color[i]) break L_check_equal;
        }

        if (back_color[3] == 1.0/* && front_color[3] == 1.0*/) {
            // rgb components are equal an their alphas are 1.0
            if (front_color != output_color && back_color != output_color) {
                for (let i = 0; i < 4; i++) output_color[i] = front_color[i];
            }
            return;
        }
    }

    if (back_color[3] <= 0) {
        // back color is transparent, return the front color
        if (front_color != output_color) {
            for (let i = 0; i < 4; i++) output_color[i] = front_color[i];
        }
        return;
    } else if (front_color[3] <= 0) {
        // front color is transparent, return the back color
        if (back_color != output_color) {
            for (let i = 0; i < 4; i++) output_color[i] = back_color[i];
        }
        return;
    }

    //
    // Color blending algorithm starts here
    //

    function normal_alpha_compose(bg_alpha, fg_alpha, cmst_a, bg_cmpt, fg_cmpt) {
        let s = fg_alpha / cmst_a;
        let t = (1 - s) * bg_cmpt;
        let u = (1 - bg_alpha) * fg_cmpt + bg_alpha * fg_cmpt/*composite color component*/;

        /* u = math2d_roundf(u); */
        return t + (s * u);
    }

    let ba = back_color[3], fa = front_color[3];

    if (ba <= 0.0) {
        output_color[0] = front_color[0];
        output_color[1] = front_color[1];
        output_color[2] = front_color[2];
        output_color[3] = fa;
        return;
    }

    let ca = fa + ba - fa * ba;
    let r = normal_alpha_compose(ba, fa, ca, back_color[0], front_color[0]);
    let g = normal_alpha_compose(ba, fa, ca, back_color[1], front_color[1]);
    let b = normal_alpha_compose(ba, fa, ca, back_color[2], front_color[2]);

    /*
    // round final color components
    r = math2d_roundf(r);
    g = math2d_roundf(g);
    b = math2d_roundf(b);
    */

    output_color[0] = r;
    output_color[1] = g;
    output_color[2] = b;
    output_color[3] = ca;
}

function math2d_color_floats_to_bytes(float_array, has_alpha) {
    let length = has_alpha ? 4 : 3;
    let shift = has_alpha ? 24 : 16;
    let argb = 0x00;

    //
    // Array disposition:
    //      rgb8    [r, g, b]
    //      rgba8   [r, g, b, a]
    //

    for (let i = 0; i < length; i++) {
        let c = (float_array[i] * 255) & 0xFF;
        argb |= c << shift;
        shift -= 8;
    }

    return argb;
}

function math2d_color_bytes_to_floats(rgba8_color, has_alpha, output_float_array) {
    let length = has_alpha ? 4 : 3;
    let shift = has_alpha ? 24 : 16;

    for (let i = 0; i < length; i++) {
        let c = (rgba8_color >>> shift) & 0xFF;
        output_float_array[i] = c / 255.0;
        shift -= 8;
    }

}

/** @deprecated */
function math2d_color_float_to_css_color(float_array, has_alpha) {
    let css_color = "#";

    function hex(array, index) {
        let c = array[index];
        c = Math.trunc(c * 255);
        c = c.toString(16);
        if (c.length < 2) c = "0".concat(c);
        return c;
    }

    let length = has_alpha ? 4 : 3;

    for (let i = 0; i < length; i++)
        css_color += hex(float_array, i);

    return css_color;
}

function math2d_bitcount(value) {
    value = value - ((value >>> 1) & 0x55555555)
    value = (value & 0x33333333) + ((value >>> 2) & 0x33333333)
    return ((value + (value >>> 4) & 0xF0F0F0F) * 0x1010101) >>> 24
}

function math2d_beats_per_minute_to_beat_per_milliseconds(bmp) {
    return 60000 / bmp;
}

function math2d_points_distance(x1, y1, x2, y2) {
    return Math.sqrt(Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2));
}

function math2d_poweroftwo_calc(dimmen) {
    let size = 2;
    while (size < 0x80000) {
        if (size >= dimmen) break;
        size *= 2;
    }

    if (size >= 0x80000) throw new Error("math2d_poweroftwo_calc() failed for: " + dimmen);
    return size;
}

function math2d_floats_are_near_equal(float1, float2) {
    return Math.abs(float1 - float2) < Number.EPSILON;
}

function math2d_float_are_near_zero(number) {
    return Math.abs(number) < Number.EPSILON;
}

function math2d_timestamp_to_string(timestamp) {
    const TIME_SECONDS = "$1ds";// 1.2s
    const TIME_MINUTES = "$im$2is";// 1m23s
    const TIME_HOURS = "$ih$2i$2i";// 1h23m45s

    if (Number.isNaN(timestamp)) return "--:--.---";

    timestamp /= 1000.0;
    let h = Math.floor(timestamp / 3600.0);
    let m = Math.floor((timestamp - (h * 3600.0)) / 60.0);
    let s = timestamp - (h * 3600.0) - (m * 60.0);

    let stringbuilder = stringbuilder_init(9);

    if (h > 0.0)
        stringbuilder_add_format(stringbuilder, TIME_HOURS, Math.trunc(h), Math.trunc(m), Math.trunc(s));
    else if (m > 0.0)
        stringbuilder_add_format(stringbuilder, TIME_MINUTES, Math.trunc(m), Math.trunc(s));
    else
        stringbuilder_add_format(stringbuilder, TIME_SECONDS, s);

    let str = stringbuilder_finalize(stringbuilder);

    return str;
}

function math2d_lerp_cubic(percent) {
    return Math.pow(percent, 3);
}

function math2d_lerp_quad(percent) {
    return Math.pow(percent, 2);
}

function math2d_lerp_expo(percent) {
    return Math.pow(2.0, 10.0 * (percent - 1));
}

function math2d_lerp_sin(percent) {
    return Math.sin(percent * MATH2D_HALF_PI);
}

File: src-desktop/kdy-e/engine/utils/math2d.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics.X86;
using System.Text;

namespace Engine.Utils;

public static class Math2D {

    public static readonly float DEG_TO_RAD = (float)(Math.PI / 180.0);
    public const int MAX_INT32 = 0x7FFFFFFF;
    public static readonly float LOG100 = MathF.Log(100);
    public const float HALF_PI = MathF.PI / 2f;

    private static readonly Random random;

    static Math2D() {
        random = new Random(Guid.NewGuid().GetHashCode());
    }

    public static float RandomFloat() {
        //return (float)(random.Next() / (double)Int32.MaxValue);
        return (float)random.NextDouble();
    }

    public static double RandomDouble() {
        return random.NextDouble();
    }

    public static float Random(float min, float max) {
        return (float)(random.NextDouble() * (max - min + 1) + min);
    }

    public static double Random(double min, double max) {
        return random.NextDouble() * (max - min + 1) + min;
    }

    public static int RandomInt(int min, int max) {
        return random.Next(min, max + 1);
    }


    public static float Lerp(float start, float end, float step) {
        return start + (end - start) * step;
    }

    public static float InverseLerp(float start, float end, float value) {
        return (value - start) / (end - start);
    }
    public static double InverseLerp(double start, double end, double value) {
        return (value - start) / (end - start);
    }

    public static float NearestDown(float value, float step) {
        return value - (value % step);
    }

    public static float CubicBezier(float offset, float point0, float point1, float point2, float point3) {
        float neg = 1 - offset;
        return (neg * neg * neg) * point0 +
            3 * offset * (neg * neg) * point1 +
            3 * (offset * offset) * neg * point2 +
            (offset * offset * offset) * point3;
    }

    public static int Clamp(int value, int min, int max) {
        if (value < min) return min;
        if (value > max) return max;
        return value;
    }

    public static float Clamp(float value, float min, float max) {
        if (value < min) return min;
        if (value > max) return max;
        return value;
    }
    public static double Clamp(double value, double min, double max) {
        if (value < min) return min;
        if (value > max) return max;
        return value;
    }

    public static void RotatePoint(float radians, ref float x, ref float y) {
        double s = Math.Sin(radians);
        double c = Math.Cos(radians);
        double X = x * c - y * s;
        double Y = x * s + y * c;

        x = (float)X;
        y = (float)Y;
    }


    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static float NormalAlphaCompose(float bg_alpha, float fg_alpha, float cmst_a, float bg_cmpt, float fg_cmpt) {
        float s = fg_alpha / cmst_a;
        float t = (1 - s) * bg_cmpt;
        float u = (1 - bg_alpha) * fg_cmpt + bg_alpha * fg_cmpt/*composite color component*/;

        /* u = math2d_roundf(u); */
        return t + (s * u);
    }

    public static void ColorBlendNormal(float[] back_color, float[] front_color, float[] output_color) {
        //
        // Avoid useless calculations if both colors are equal, also both alphas must be 1.0
        //

        for (int i = 0 ; i < 4 ; i++) {
            if (back_color[i] != front_color[i]) goto L_stop_checking_equal;
        }

        if (back_color[3] == 1.0/* && front_color[3] == 1.0*/) {
            // rgb components are equal an their alphas are 1.0
            if (front_color != output_color && back_color != output_color) {
                for (int i = 0 ; i < 4 ; i++) output_color[i] = front_color[i];
            }
            return;
        }

L_stop_checking_equal:
        if (back_color[3] <= 0) {
            // back color is transparent, return the front color
            if (front_color != output_color) {
                for (int i = 0 ; i < 4 ; i++) output_color[i] = front_color[i];
            }
            return;
        } else if (front_color[3] <= 0) {
            // front color is transparent, return the back color
            if (back_color != output_color) {
                for (int i = 0 ; i < 4 ; i++) output_color[i] = back_color[i];
            }
            return;
        }

        //
        // Color blending algorithm starts here
        //


        float ba = back_color[3], fa = front_color[3];

        if (ba <= 0.0) {
            output_color[0] = front_color[0];
            output_color[1] = front_color[1];
            output_color[2] = front_color[2];
            output_color[3] = fa;
            return;
        }

        float ca = fa + ba - fa * ba;
        float r = NormalAlphaCompose(ba, fa, ca, back_color[0], front_color[0]);
        float g = NormalAlphaCompose(ba, fa, ca, back_color[1], front_color[1]);
        float b = NormalAlphaCompose(ba, fa, ca, back_color[2], front_color[2]);

        /*
        // round final color components
        r = math2d_roundf(r);
        g = math2d_roundf(g);
        b = math2d_roundf(b);
        */

        output_color[0] = r;
        output_color[1] = g;
        output_color[2] = b;
        output_color[3] = ca;
    }

    public static uint ColorFloatsToBytes(float[] float_array, bool has_alpha) {
        int length = has_alpha ? 4 : 3;
        int shift = has_alpha ? 24 : 16;
        uint argb = 0x00;

        //
        // Array disposition:
        //      rgb8    [r, g, b]
        //      rgba8   [r, g, b, a]
        //

        for (int i = 0 ; i < length ; i++) {
            uint c = ((uint)(float_array[i] * 255)) & 0xFFU;
            argb |= c << shift;
            shift -= 8;
        }

        return argb;
    }

    public static void ColorBytesToFloats(uint rgba8_color, bool has_alpha, float[] output_float_array) {
        int length = has_alpha ? 4 : 3;
        int shift = has_alpha ? 24 : 16;

        for (int i = 0 ; i < length ; i++) {
            uint c = (rgba8_color >> shift) & 0xFF;
            output_float_array[i] = c / 255.0f;
            shift -= 8;
        }

    }

    public static uint BitCount(uint value) {
        if (Popcnt.IsSupported) {
            return Popcnt.PopCount(value);
        }

        // POPCNT .Net Impl
        value = value - ((value >> 1) & 0x55555555);
        value = (value & 0x33333333) + ((value >> 2) & 0x33333333);
        return ((value + (value >> 4) & 0xF0F0F0F) * 0x1010101) >> 24;
    }

    public static float BeatsPerMinuteToBeatPerMilliseconds(float bpm) {
        return 60000.0f / bpm;
    }

    public static float PointsDistance(float x1, float y1, float x2, float y2) {
        return MathF.Sqrt(MathF.Pow(x1 - x2, 2f) + MathF.Pow(y1 - y2, 2f));
    }

    public static int PowerOfTwoCalc(int dimmen) {
        int size = 2;
        while (size < 0x80000) {
            if (size >= dimmen) break;
            size *= 2;
        }

        if (size >= 0x80000) throw new ArgumentOutOfRangeException("math2d_poweroftwo_calc() failed for: " + dimmen);
        return size;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool FloatsAreNearEqual(float float1, float float2) {
        return Math.Abs(float1 - float2) < Single.Epsilon;
    }

    public static bool FloatsAreNearZero(float number) {
        return Math.Abs(number) < Single.Epsilon;
    }

    public static string TimestampToString(double timestamp) {
        const string TIME_SECONDS = "$1ds";// 1.2s
        const string TIME_MINUTES = "$im$2is";// 1m23s
        const string TIME_HOURS = "$ih$2i$2i";// 1h23m45s

        if (Double.IsNaN(timestamp)) return "--:--.---";

        timestamp /= 1000.0;
        double h = Math.Floor(timestamp / 3600.0);
        double m = Math.Floor((timestamp - (h * 3600.0)) / 60.0);
        double s = timestamp - (h * 3600.0) - (m * 60.0);

        StringBuilder stringbuilder = new StringBuilder(9);

        if (h > 0.0)
            stringbuilder.AddFormatKDY(TIME_HOURS, (int)h, (int)m, (int)s);
        else if (m > 0.0)
            stringbuilder.AddFormatKDY(TIME_MINUTES, (int)m, (int)s);
        else
            stringbuilder.AddFormatKDY(TIME_SECONDS, s);

        string str = stringbuilder.GetCopyKDY();
        //stringbuilder.Destroy();

        return str;
    }

    public static float LerpCubic(float percent) {
        return MathF.Pow(percent, 3f);
    }

    public static float LerpQuad(float percent) {
        return MathF.Pow(percent, 2f);
    }

    public static float LerpExpo(float percent) {
        return MathF.Pow(2f, 10f * (percent - 1f));
    }

    public static float LerpSin(float percent) {
        return MathF.Sin(percent * Math2D.HALF_PI);
    }


}